cli/lsp/documents.rs

```diff
@@ -984,63 +1059,38 @@ impl DocumentsInner {
     None
   }
 
-  fn resolve_remote_specifier(
-    &self,
-    specifier: &ModuleSpecifier,
-    redirect_limit: usize,
-  ) -> Option<ModuleSpecifier> {
-    let cache_filename = self.cache.get_cache_filename(specifier)?;
-    if redirect_limit > 0 && cache_filename.is_file() {
-      let headers = http_cache::Metadata::read(&cache_filename)
-        .ok()
-        .map(|m| m.headers)?;
-      if let Some(location) = headers.get("location") {
-        let redirect =
-          deno_core::resolve_import(location, specifier.as_str()).ok()?;
-        self.resolve_remote_specifier(&redirect, redirect_limit - 1)
-      } else {
-        Some(specifier.clone())
-      }
-    } else {
-      None
-    }
-  }
-
-  fn resolve_specifier(
-    &mut self,
-    specifier: &ModuleSpecifier,
-  ) -> Option<ModuleSpecifier> {
-    let scheme = specifier.scheme();
-    if !SUPPORTED_SCHEMES.contains(&scheme) {
-      return None;
-    }
-
-    if scheme == "data" || scheme == "blob" || scheme == "file" {
-      Some(specifier.clone())
-    } else if let Some(specifier) = self.redirects.get(specifier) {
-      Some(specifier.clone())
-    } else {
-      let redirect = self.resolve_remote_specifier(specifier, 10)?;
-      self.redirects.insert(specifier.clone(), redirect.clone());
-      Some(redirect)
-    }
-  }
-
   fn set_location(&mut self, location: PathBuf) {
     // TODO update resolved dependencies?
     self.cache = HttpCache::new(&location);
+    self.specifier_resolver = Arc::new(SpecifierResolver::new(&location));
     self.dirty = true;
   }
 
-  fn set_navigation_tree(
+  fn try_set_navigation_tree(
     &mut self,
     specifier: &ModuleSpecifier,
+    script_version: &str,
     navigation_tree: Arc<tsc::NavigationTree>,
   ) -> Result<(), AnyError> {
-    let doc = self.docs.get_mut(specifier).ok_or_else(|| {
-      custom_error("NotFound", format!("Specifier not found {}", specifier))
-    })?;
-    *doc = doc.with_navigation_tree(navigation_tree);
+    if let Some(doc) = self.open_docs.get_mut(specifier) {
```

Comment:
does this correctly handle situations where there is already a cached navigation tree that is now invalid? (e.g. there is an invalid nav tree already cached, we go to calculate a new one, but the document has changed while calculating it meaning we don't update the old old nav tree, meaning the cached one is even more out of date)

Also, does it correctly handle situations where we don't get into an infinite loop of constantly trying to update the nav tree?

Reply:
It does feel overall unrelated to the _core_ of this PR.

Reply:
Ok... I was looking at code in vscode recently and they have effectively a debounce for this:

```ts
function waitForDocumentChangesToEnd(document: vscode.TextDocument) {
	let version = document.version;
	return new Promise<void>((s) => {
		const iv = setInterval(_ => {
			if (document.version === version) {
				clearInterval(iv);
				s();
			}
			version = document.version;
		}, 400);
	});
}
```

We put a debounce in calculating diagnostics as well (basically debouncing the calculation of them because they can occur async).  It isn't the scope of this PR, but we might want to consider this.  Both the semantic tokens and the nav tree are the costly tsc calls and they are the things that can make it feel sluggish.

Reply:
It might be good to do in the meantime until we can do #11032. I think very likely that one will make it fast enough that it won't matter (and we'll be able to get rid of this caching code too).